components/sync15/src/collection_state.rs

```diff
@@ -0,0 +1,174 @@
+    pub config: InfoConfiguration,
+    // initially from meta/global, updated after an xius POST/PUT.
+    pub last_modified: ServerTimestamp,
+    pub key: KeyBundle,
+}
+
+pub enum LocalCollectionState {
+    Unknown {
+        global_sync_id: Option<String>,
+        store_sync_id: Option<String>,
+    },
+    Declined,
+    GlobalSyncIdChanged,
+    StoreSyncIdChanged {
+        global_sync_id: String,
+    },
+    Ready {
+        key: KeyBundle,
+        global_sync_id: String,
+        store_sync_id: String,
+    },
+}
+
+pub struct LocalCollectionStateMachine<'state> {
+    global_state: &'state GlobalState,
+}
+
+impl<'state> LocalCollectionStateMachine<'state> {
+    fn advance(
+        &self,
+        from: LocalCollectionState,
+        store: &Store,
+    ) -> error::Result<LocalCollectionState> {
+        let name = &store.collection_name().to_string();
+        let meta_global = &self.global_state.global;
+        match from {
+            LocalCollectionState::Unknown {
+                global_sync_id: maybe_gsid,
+                store_sync_id: maybe_ssid,
+            } => {
+                if meta_global.declined.contains(name) {
+                    return Ok(LocalCollectionState::Declined);
+                }
+                let global_sync_id = meta_global.sync_id.clone();
+                if maybe_gsid.map(|id| id != global_sync_id).unwrap_or(true) {
+                    return Ok(LocalCollectionState::GlobalSyncIdChanged);
+                }
+                let store_sync_id = meta_global.engines[name].sync_id.clone();
+                if maybe_ssid.map(|id| id != store_sync_id).unwrap_or(true) {
+                    return Ok(LocalCollectionState::StoreSyncIdChanged { global_sync_id });
+                }
+                Ok(LocalCollectionState::Ready {
+                    key: self.global_state.keys.default.clone(),
+                    global_sync_id,
+                    store_sync_id,
+                })
+            }
+
+            LocalCollectionState::Declined => unreachable!("can't advance from declined"),
+
+            LocalCollectionState::GlobalSyncIdChanged
+            | LocalCollectionState::StoreSyncIdChanged { .. } => {
+                // we treat either guid changing the same way - we already have
+                // a new meta/global with new IDs - so grab them both and advance.
+                let global_sync_id = meta_global.sync_id.clone();
+                let store_sync_id = meta_global.engines[name].sync_id.clone(); // xxx - how does `[name]` not potentially fail?
+                store.reset(&global_sync_id, &store_sync_id)?;
+                Ok(LocalCollectionState::Unknown {
+                    global_sync_id: Some(global_sync_id),
+                    store_sync_id: Some(store_sync_id),
+                })
+            }
+
+            LocalCollectionState::Ready { .. } => unreachable!("can't advance from ready"),
+        }
+    }
+
+    fn run_and_run_as_farst_as_you_can(
```

Comment:
s/farst/fast/.

Also, lol.

Reply:
(lol - I invented a new word - I wanted the state machine to run as *far* as it can, but a gingerbread man wants to run as *fast* as he can ;)

I'll come up with a better name for both ;)

Reply:
Ah, the old frecency approach.